### PR TITLE
fix: refactor and fix bulk attribute handling

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

Fixes: CAL-4666

Created: https://github.com/calcom/cal.com/issues/18230 this was broken outside of this PR. We should fix it at some point

## Mandatory Tasks (DO NOT REMOVE)

- [X] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [X] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Assign different attribute types to users and see expected outcome (you will need to manually refresh issue created above to fix optimistic UI here)
Mainly check single select can only ever get one value when updating

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/prisma/index.ts [hunk 1]
```diff
@@ -81,6 +81,15 @@ if (process.env.NODE_ENV !== "production") {
...
 type PrismaClientWithExtensions = typeof prismaWithClientExtensions;
 export type PrismaClient = PrismaClientWithExtensions;
+
+type OmitPrismaClient = Omit<
+  PrismaClient,
+  "$connect" | "$disconnect" | "$on" | "$transaction" | "$use" | "$extends"
+>;
+
+// we cant pass tx to functions as types miss match since we have a custom prisma client https://github.com/prisma/prisma/discussions/20924#discussioncomment-10077649
+export type PrismaTransaction = OmitPrismaClient;
```
Comment (sean-brydon) [inline packages/prisma/index.ts:91]: We cant just use Pirsma.TransactionClient for this as we have an extended prisma

File: packages/trpc/server/routers/viewer/attributes/attributeUtils.ts
Comment (sean-brydon) [inline packages/trpc/server/routers/viewer/attributes/attributeUtils.ts:1]: Extracted to UTILS for readability and allows us to easily test all the madness that happens here

File: packages/trpc/server/routers/viewer/attributes/bulkAssignAttributes.handler.ts [hunk 2]
```diff
@@ -23,6 +23,9 @@ const bulkAssignAttributesHandler = async ({ input, ctx }: GetOptions) => {
     });
   }
 
+  // Create a map of attribute types for quick lookup
+  const attributeTypes = new Map<string, string>();
```
Comment (sean-brydon) [inline packages/trpc/server/routers/viewer/attributes/bulkAssignAttributes.handler.ts:27]: This was much more efficient than doing it per iteration as multiple users can be assigned the same attribute

**Review by eunjae-lee — APPROVED** [review]
the code looks good and it's working well

**Pull request merged**